kura/org.eclipse.kura.linux.clock/src/main/java/org/eclipse/kura/linux/clock/NtsClockSyncProvider.java

```diff
@@ -0,0 +1,254 @@
+/*******************************************************************************
+ * Copyright (c) 2021 Eurotech and/or its affiliates and others
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *  Eurotech
+ *******************************************************************************/
+package org.eclipse.kura.linux.clock;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.Date;
+import java.util.Map;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.kura.KuraErrorCode;
+import org.eclipse.kura.KuraException;
+import org.eclipse.kura.executor.Command;
+import org.eclipse.kura.executor.CommandExecutorService;
+import org.eclipse.kura.executor.CommandStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.annotations.SerializedName;
+
+public class NtsClockSyncProvider implements ClockSyncProvider {
+
+    private static final String NTS_PROVIDER = "chronyd";
+
+    private static final Logger logger = LoggerFactory.getLogger(NtsClockSyncProvider.class);
+
+    private final CommandExecutorService executorService;
+    private ScheduledExecutorService schedulerExecutor;
+
+    private Map<String, Object> properties;
+    private ClockSyncListener listener;
+
+    private Path chronyConfigLocation;
+    private String ntsConfig;
+
+    private Date lastSyncValue;
+    private long lastOffset;
+
+    private Gson gson;
+
+    private long lastSyncTime;
+
+    public NtsClockSyncProvider(CommandExecutorService service) {
+        this.executorService = service;
+    }
+
+    @Override
+    public void init(Map<String, Object> properties, ClockSyncListener listener) throws KuraException {
+        this.properties = properties;
+        this.listener = listener;
+
+        this.gson = new Gson();
+
+        readProperties();
+        writeConfiguration();
+    }
+
+    private void writeConfiguration() {
+
+        if (this.ntsConfig != null && !this.ntsConfig.isEmpty()) {
+            try {
+                logger.info("Saving previous chrony configuration file at {}", this.chronyConfigLocation);
+                Files.copy(this.chronyConfigLocation, Paths.get(this.chronyConfigLocation.toString() + ".bak"));
```

Comment:
This seems to fail if the destination file exists, we should also check what happens if the source file does not exist